### PR TITLE
Added the SCP-079 Recontaining event

### PR DESCRIPTION
**Pull request by FoxWorn3365**

### CODE REVIEW

File: EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            int index = 0;
+            List<CodeInstruction> newInstructions = ListPool<CodeInstruction>.Pool.Get(instructions);
+
+            LocalBuilder ev = generator.DeclareLocal(typeof(RecontainingEventArgs));
+
+            Label returnLabel = generator.DefineLabel();
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // RecontainingEventArgs ev = new(ReferenceHub "attacker")
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:43]:
```suggestion
                // RecontainingEventArgs ev = new(this._activatorGlass)
```

File: EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Ldfld, PropertyGetter(typeof(Scp079Recontainer), nameof(Scp079Recontainer._activatorGlass))),
+                new(OpCodes.Newobj, GetDeclaredConstructors(typeof(RecontainingEventArgs))[1]),
+                new(OpCodes.Starg_S, ev.LocalIndex),
+
+                // Call event
```
Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:49]: ev.OnRecontaining

Comment (louis1706) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:49]:
```suggestion
                // Scp079.OnRecontaining(ev)
```

**Comment by VALERA771** [comment]
Transpiler tested?

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    /// </summary>
+    public class RecontainingEventArgs : IDeniableEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RecontainingEventArgs" /> class.
+        /// </summary>
+        /// <param name="recontainer">The <see cref="BreakableWindow"/> instance.</param>
+        public RecontainingEventArgs(BreakableWindow recontainer)
+        {
+            Recontainer = Player.Get(recontainer?.LastAttacker.Hub);
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:24]: recontainer can't be null but LastAttacker could be also please make IsAutomatic with recontainer.LastAttacker.IsSet

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs (hunk 3 above)
Comment (FoxWorn3365) [reply EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:24]: LastAttacker.Hub can't be null.
Anyways if recontainer.LastAttacker is null Player.Get() will return a null so it's the same thing

Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:24]:
```suggestion
            Recontainer = Player.Get(recontainer.LastAttacker.Hub);
            IsAutomatic = recontainer.LastAttacker.IsSet;
```

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        /// </summary>
+        public Player Recontainer { get; }
+
+        /// <inheritdoc/>
+        public bool IsAllowed { get; set; } = true;
+
+        /// <summary>
+        /// Gets a value indicating whether or not the recontained has been made automatically or by triggering the proccess.
+        /// </summary>
+        public bool IsAutomatic => Recontainer is null;
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:39]:
```suggestion
        public bool IsAutomatic { get; }
```

**Review by louis1706 — CHANGES_REQUESTED** [review]
Final requests

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs (hunk 3 above)
Comment (louis1706) [reply EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:24]: Could contain error for property name i have made it on phone

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace Exiled.Events.EventArgs.Scp079
+{
+    using Exiled.API.Features;
+    using Exiled.Events.EventArgs.Interfaces;
+
+    /// <summary>
+    /// Contains information before SCP-079 gets recontained.
+    /// </summary>
+    public class RecontainingEventArgs : IDeniableEvent
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:16]: Add IPlayerEvent

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs (hunk 5 above)
Comment (FoxWorn3365) [reply EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:16]: I'll do it tomorrow

**Review by Misfiy — APPROVED** [review]

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs (hunk 5 above)
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:16]:
```suggestion
    public class RecontainingEventArgs : IDeniableEvent, IPlayerEvent
```

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            Recontainer = Player.Get(recontainer.LastAttacker.Hub);
+            IsAutomatic = recontainer.LastAttacker.IsSet;
+        }
+
+        /// <summary>
+        /// Gets the Player that started the recontainment process.<br></br>
+        /// Can be null if <see cref="IsAutomatic"/> is true.
+        /// </summary>
+        public Player Recontainer { get; }
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:32]:
```suggestion
        public Player Player { get; }
```

File: EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    /// </summary>
+    public class RecontainingEventArgs : IDeniableEvent
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RecontainingEventArgs" /> class.
+        /// </summary>
+        /// <param name="recontainer">The <see cref="BreakableWindow"/> instance.</param>
+        public RecontainingEventArgs(BreakableWindow recontainer)
+        {
+            Recontainer = Player.Get(recontainer.LastAttacker.Hub);
```
Comment (louis1706) [inline EXILED/Exiled.Events/EventArgs/Scp079/RecontainingEventArgs.cs:24]:
```suggestion
            Player = Player.Get(recontainer.LastAttacker.Hub);
```

**Review by louis1706 — APPROVED** [review]

**Review by VALERA771 — APPROVED** [review]

File: EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+            Label returnLabel = generator.DefineLabel();
+
+            newInstructions.InsertRange(index, new CodeInstruction[]
+            {
+                // RecontainingEventArgs ev = new(this._activatorGlass)
+                new(OpCodes.Ldarg_0),
+                new(OpCodes.Ldfld, PropertyGetter(typeof(Scp079Recontainer), nameof(Scp079Recontainer._activatorGlass))),
+                new(OpCodes.Newobj, GetDeclaredConstructors(typeof(RecontainingEventArgs))[0]),
+                new(OpCodes.Starg_S, ev.LocalIndex),
```
Comment (VALERA771) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:47]:
```suggestion
                new(OpCodes.Stloc_S, ev.LocalIndex),
```

Are you sure that starg?

File: EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+                // Scp079.OnRecontaining(ev)
+                new(OpCodes.Ldarg_S, ev.LocalIndex),
```
Comment (VALERA771) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:50]:
```suggestion
                new(OpCodes.Ldloc_S, ev.LocalIndex),
```

File: EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+                new(OpCodes.Call, Method(typeof(Scp079), nameof(Scp079.OnRecontaining))),
+
+                // if (!ev.IsAllowed) return;
+                new(OpCodes.Ldarg_S, ev.LocalIndex),
```
Comment (VALERA771) [inline EXILED/Exiled.Events/Patches/Events/Scp079/Recontaining.cs:54]:
```suggestion
                new(OpCodes.Ldloc_S, ev.LocalIndex),
```

**Review by VALERA771 — CHANGES_REQUESTED** [review]
Do not pass event value as a function argument

**Review by VALERA771 — APPROVED** [review]

**Review by Misfiy — APPROVED** [review]

**Pull request merged**